state/protocol/badger/snapshot_test.go

```diff
@@ -301,14 +300,125 @@ func TestSealingSegment(t *testing.T) {
 			require.NoError(t, err)
 
 			// sealing segment should be [B2, B3, B4]
-			assert.Len(t, segment, 3)
-			assert.Equal(t, block2.ID(), segment[0].ID())
-			assert.Equal(t, block3.ID(), segment[1].ID())
-			assert.Equal(t, block4.ID(), segment[2].ID())
+			assert.Len(t, segment.Blocks, 3)
+			assert.Equal(t, block2.ID(), segment.Blocks[0].ID())
+			assert.Equal(t, block3.ID(), segment.Blocks[1].ID())
+			assert.Equal(t, block4.ID(), segment.Blocks[2].ID())
 		})
 	})
-}
 
+	// test sealing segment where you have a chain that is 5 blocks long and the block 5 has a seal for block 2.
+	// block 2 also contains a receipt but no result.
+	// root -> B1[Result_A, Receipt_A_1] -> B2[Result_B, Receipt_B, Receipt_A_2] -> B3 -> B4 -> B5 (Seal_B2)
+	t.Run("sealing segment with 4 blocks and 1 execution result decoupled", func(t *testing.T) {
+		util.RunWithFollowerProtocolState(t, rootSnapshot, func(db *badger.DB, state *bprotocol.FollowerState) {
+			// simulate scenario where execution result is missing from block payload
+			// SealingSegment() should get result from results db and store it on ExecutionReceipts
+			// field on SealingSegment
+			resultA := unittest.ExecutionResultFixture()
+			receiptA1 := unittest.ExecutionReceiptFixture(unittest.WithResult(resultA))
+			receiptA2 := unittest.ExecutionReceiptFixture(unittest.WithResult(resultA))
+
+			// receipt b also contains result b
+			receiptB := unittest.ExecutionReceiptFixture()
+
+			block1 := unittest.BlockWithParentFixture(head)
+			block1.SetPayload(unittest.PayloadFixture(unittest.WithReceipts(receiptA1)))
+
+			block2 := unittest.BlockWithParentFixture(block1.Header)
+			block2.SetPayload(unittest.PayloadFixture(unittest.WithReceipts(receiptB), unittest.WithReceiptsAndNoResults(receiptA2)))
+
+			block3 := unittest.BlockWithParentFixture(block2.Header)
+			block4 := unittest.BlockWithParentFixture(block3.Header)
+
+			block5 := unittest.BlockWithParentFixture(block4.Header)
+			receiptForSeal, seal := unittest.ReceiptAndSealForBlock(&block2)
+			block5.SetPayload(unittest.PayloadFixture(unittest.WithSeals(seal), unittest.WithReceipts(receiptForSeal)))
+
+			err := state.Extend(context.Background(), &block1)
+			require.NoError(t, err)
+			err = state.Extend(context.Background(), &block2)
+			require.NoError(t, err)
+			err = state.Extend(context.Background(), &block3)
+			require.NoError(t, err)
+			err = state.Extend(context.Background(), &block4)
+			require.NoError(t, err)
+			err = state.Extend(context.Background(), &block5)
+			require.NoError(t, err)
+
+			segment, err := state.AtBlockID(block5.ID()).SealingSegment()
+			require.NoError(t, err)
+
+			require.Len(t, segment.Blocks, 4)
+
```

Comment:
Is it possible to add such check?
```suggestion
			require.Len(t, segment.Blocks, 4)
			require.Equal(t, []*flow.Block{&block2, &block3, &block4, &block5}, segment.Blocks)
```

Reply:
https://github.com/onflow/flow-go/pull/1533/commits/00a77e58055c9b7fbc3a8d569aa91866610f5646